kernel/vmbytegen.ml

```diff
@@ -799,37 +813,43 @@ let rec compile_lam env cenv lam sz cont =
 
 and compile_get_global env cenv (kn,u) sz cont =
   let () = set_max_stack_size cenv sz in
-  if Univ.Instance.is_empty u then
+  if UVars.Instance.is_empty u then
     Kgetglobal kn :: cont
   else
     comp_app (fun _ _ _ cont -> Kgetglobal kn :: cont)
       (compile_instance env) cenv () [|u|] sz cont
 
-and compile_instance env cenv u0 sz cont =
+and compile_instance env cenv u sz cont =
   let () = set_max_stack_size cenv sz in
-  let u = Univ.Instance.to_array u0 in
-  let len = Array.length u in
-  let is_id i l = match Univ.Level.var_index l with
-  | None -> false
-  | Some j -> Int.equal i j
-  in
-  if Int.equal env.uinst_len len && Array.for_all_i is_id 0 u then
+  if is_toplevel_inst env u then
     (* Optimization: do not reallocate the same instance *)
     pos_instance cenv sz :: cont
-  else if Array.for_all (fun l -> Option.is_empty (Univ.Level.var_index l)) u then
+  else if is_closed_inst u then
     (* Optimization: allocate closed instances globally *)
-    compile_structured_constant cenv (Const_univ_instance u0) sz cont
+    compile_structured_constant cenv (Const_univ_instance u) sz cont
   else
-    let () = set_max_stack_size cenv (sz + len - 1) in
+    let qs, us = UVars.Instance.to_array u in
+    let qlen, ulen = UVars.Instance.length u in
+    let () = set_max_stack_size cenv (sz + qlen + ulen - 1) in
```

Comment:
This is a gross overestimation, since the blocks are constructed one after the other and thus occupy roughly the same stack slots. Now that you have a `comp_array` function, it would be better to just do the following computation there:
```ocaml
set_max_stack_size cenv (sz + Array.length vs - 1);
```

Reply:
Should we also add a +1 somewhere since we push the univ array?

Reply:
The +1 is already part of the precise size tracking when doing
```ocaml
let cont = comp_array comp_qual cenv qs (sz+1) cont in
```

Reply:
OK, changed